File: Contracts/src/main/java/org/cubexmc/contract/model/PayoutRecipient.java
package org.cubexmc.contract.model;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public final class PayoutRecipient {
    public enum Kind {
        PARTICIPANT,
        SYSTEM_SINK,
        ARBITER
    }

    private final Kind kind;
    private final ParticipantRole role;

    private PayoutRecipient(Kind kind, ParticipantRole role) {
        this.kind = kind;
        this.role = role;
    }

    public static PayoutRecipient participant(ParticipantRole role) {
        return new PayoutRecipient(Kind.PARTICIPANT, role);
    }

    public static PayoutRecipient systemSink() {
        return new PayoutRecipient(Kind.SYSTEM_SINK, null);
    }

    public static PayoutRecipient arbiter() {
        return new PayoutRecipient(Kind.ARBITER, null);
    }

    public Kind kind() {
        return kind;
    }

    public ParticipantRole role() {
        return role;
    }

    public Map<String, Object> toMap() {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("kind", kind.name());
        if (role != null) {
            map.put("role", role.name());
        }
        return map;
    }

    public static PayoutRecipient fromMap(Map<?, ?> map) {
        Kind kind = Kind.valueOf(Objects.toString(map.get("kind"), "SYSTEM_SINK"));
        if (kind == Kind.PARTICIPANT) {
            return participant(ParticipantRole.valueOf(Objects.toString(map.get("role"), "OWNER")));
        }
        if (kind == Kind.ARBITER) {
            return arbiter();
        }
        return systemSink();
    }
}

File: Contract/src/main/java/org/cubexmc/contract/model/PayoutRecipient.kt
package org.cubexmc.contract.model

import java.util.LinkedHashMap
import java.util.Objects

class PayoutRecipient private constructor(
    private val kind: Kind,
    private val role: ParticipantRole?,
) {
    enum class Kind {
        PARTICIPANT,
        SYSTEM_SINK,
        ARBITER,
    }

    fun kind(): Kind = kind

    fun role(): ParticipantRole? = role

    fun toMap(): Map<String, Any?> {
        val map = LinkedHashMap<String, Any?>()
        map["kind"] = kind.name
        if (role != null) {
            map["role"] = role.name
        }
        return map
    }

    companion object {
        @JvmStatic
        fun participant(role: ParticipantRole): PayoutRecipient = PayoutRecipient(Kind.PARTICIPANT, role)

        @JvmStatic
        fun systemSink(): PayoutRecipient = PayoutRecipient(Kind.SYSTEM_SINK, null)

        @JvmStatic
        fun arbiter(): PayoutRecipient = PayoutRecipient(Kind.ARBITER, null)

        @JvmStatic
        fun fromMap(map: Map<*, *>): PayoutRecipient {
            val kind = Kind.valueOf(Objects.toString(map["kind"], "SYSTEM_SINK"))
            if (kind == Kind.PARTICIPANT) {
                return participant(ParticipantRole.valueOf(Objects.toString(map["role"], "OWNER")))
            }
            if (kind == Kind.ARBITER) {
                return arbiter()
            }
            return systemSink()
        }
    }
}
